### ISSUE TITLE
Refactor tunnel daemon (IPC service)

### ISSUE THREAD

**Issue by ReactorScram** [CloudFire-LLC/cloudfire#6913]
I've had a few failed attempts to refactor this, and the current design is getting in the way:

- The prototype GTK Client depends on all of Tauri's dependencies, because the tunnel daemon is a target of the Tauri Client's crate, so it depends on the Tauri Client being built.
- The Tauri Client indirectly depends on things like `str0m` that aren't part of the GUI process at all, because it reuses code from the tunnel daemon

Goals:

- Types like `ClientMsg` and `ServerMsg` for IPC should not drag in all of connlib
- The GTK Client should not drag in the Tauri Client, nor connlib
- The tunnel daemon should not be part of the Headless Client's crate, but it can depend on the Headless Client.
- The `known_dirs` module needs to be available to the Headless Client (for device ID), the tunnel daemon, and the GUI process, or it could be split up somehow

Considering these target binaries:

- Headless Client
- Tunnel Daemon (IPC service)
- GTK GUI
- Tauri GUI

We have these combinations of needs:

- Needed by Headless Client and Tunnel Daemon, e.g. connlib
- Needed by Tunnel Daemon, GTK GUI, and Tauri GUI, e.g. IPC messages
- Needed by GTK GUI and Tauri GUI, e.g. GUI controller module

**Comment by thomaseizinger** [comment]
I ran into this yesterday too. I wanted to inline `connlib-shared` - which just contains message types at the moment - into `firezone-tunnel` but that would have coupled the gui-client directly to firezone-tunnel because it depends on the `callbacks::ResourceDescription` types.

So that will need to be decoupled too. Although I did think afterwards, perhaps it does make sense to have a crate for all the message types / the domain model? Perhaps that crate should be called `connlib-model`?

**Comment by ReactorScram** [comment]
Yeah, maybe. I'm not sure if I'll need new crates for this tunnel daemon thing, but I suspect there might be at least one.

**Comment by thomaseizinger** [comment]
I think it makes sense to have a crate for simple types like `ResourceId`. You'll have to handle that in the GUI and otherwise you have to duplicate them.

**Comment by thomaseizinger** [comment]
I attempted this in https://github.com/firezone/firezone/pull/6919, I think it worked out well.
